tests/test_select.py

```diff
@@ -60,22 +61,41 @@ def test_select(
 @pytest.mark.parametrize("struct", [False, True])
 @pytest.mark.parametrize("default", [10, None])
 @pytest.mark.parametrize(
-    "cfg, key",
+    ["cfg", "key", "exc_no_default"],
     [
-        pytest.param({}, "not_found", id="empty"),
-        pytest.param({"missing": "???"}, "missing", id="missing"),
-        pytest.param({"inter": "${bad_key}"}, "inter", id="inter_bad_key"),
+        pytest.param({}, "not_found", None, id="empty"),
+        pytest.param({"missing": "???"}, "missing", None, id="missing"),
+        pytest.param(
+            {"int": 0},
+            "int.y",
+            pytest.raises(
+                ConfigKeyError,
+                match=re.escape(
+                    "Error trying to access int.y: node `int` is not a container "
+                    "and thus cannot contain `y`"
+                ),
+            ),
+            id="non_container",
+        ),
     ],
 )
 def test_select_default(
     cfg: Any,
     struct: bool,
     key: Any,
+    exc_no_default: Any,
     default: Any,
 ) -> None:
     cfg = _ensure_container(cfg)
     OmegaConf.set_struct(cfg, struct)
     assert OmegaConf.select(cfg, key, default=default) == default
+    if exc_no_default is None:
+        # Not providing a default value is expected to work and return `None`.
+        assert OmegaConf.select(cfg, key) is None
+    else:
+        # Not providing a default value is expected to raise an exception.
+        with exc_no_default:
+            OmegaConf.select(cfg, key)
```

Comment:
I prefer splitting tests into positive and negative cases in scenarios like this one.
It makes the tests significantly simpler.

Reply:
Done in two steps:

First, c396527 addresses this specific issue. It's easier to see the diff against master after applying this commit, which looks like this:

```diff
@pytest.mark.parametrize("struct", [True, False, None])
@@ -60,11 +61,11 @@ def test_select(
 @pytest.mark.parametrize("struct", [False, True])
 @pytest.mark.parametrize("default", [10, None])
 @pytest.mark.parametrize(
-    "cfg, key",
+    ("cfg", "key"),
     [
         pytest.param({}, "not_found", id="empty"),
         pytest.param({"missing": "???"}, "missing", id="missing"),
-        pytest.param({"inter": "${bad_key}"}, "inter", id="inter_bad_key"),
+        pytest.param({"int": 0}, "int.y", id="non_container"),
     ],
 )
 def test_select_default(
@@ -78,6 +79,31 @@ def test_select_default(
     assert OmegaConf.select(cfg, key, default=default) == default


+@pytest.mark.parametrize("struct", [False, True])
+@pytest.mark.parametrize(
+    ("cfg", "key", "exc"),
+    [
+        pytest.param(
+            {"int": 0},
+            "int.y",
+            pytest.raises(
+                ConfigKeyError,
+                match=re.escape(
+                    "Error trying to access int.y: node `int` is not a container "
+                    "and thus cannot contain `y`"
+                ),
+            ),
+            id="non_container",
+        ),
+    ],
+)
+def test_select_error(cfg: Any, key: Any, exc: Any, struct: bool) -> None:
+    cfg = _ensure_container(cfg)
+    OmegaConf.set_struct(cfg, struct)
+    with exc:
+        OmegaConf.select(cfg, key)
+
```

Second, in ef61b9a I refactored `test_select.py` so that all tests are within a common class parameterized on the `struct` flag (otherwise it seemed a bit random)